// eslint-plugin-formatjs/dist/rules/blacklist-elements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
const intl_messageformat_parser_1 = require("intl-messageformat-parser");
class BlacklistElement extends Error {
    constructor(type) {
        super();
        this.message = `${type} element is blacklisted`;
    }
}
var Element;
(function (Element) {
    Element["literal"] = "literal";
    Element["argument"] = "argument";
    Element["number"] = "number";
    Element["date"] = "date";
    Element["time"] = "time";
    Element["select"] = "select";
    Element["selectordinal"] = "selectordinal";
    Element["plural"] = "plural";
})(Element || (Element = {}));
function verifyAst(blacklist, ast) {
    for (const el of ast) {
        if (intl_messageformat_parser_1.isLiteralElement(el) && blacklist.includes(Element.literal)) {
            throw new BlacklistElement(Element.literal);
        }
        if (intl_messageformat_parser_1.isArgumentElement(el) && blacklist.includes(Element.argument)) {
            throw new BlacklistElement(Element.argument);
        }
        if (intl_messageformat_parser_1.isNumberElement(el) && blacklist.includes(Element.number)) {
            throw new BlacklistElement(Element.number);
        }
        if (intl_messageformat_parser_1.isDateElement(el) && blacklist.includes(Element.date)) {
            throw new BlacklistElement(Element.date);
        }
        if (intl_messageformat_parser_1.isTimeElement(el) && blacklist.includes(Element.time)) {
            throw new BlacklistElement(Element.time);
        }
        if (intl_messageformat_parser_1.isSelectElement(el) && blacklist.includes(Element.select)) {
            throw new BlacklistElement(Element.select);
        }
        if (intl_messageformat_parser_1.isPluralElement(el)) {
            if (blacklist.includes(Element.plural)) {
                throw new BlacklistElement(Element.argument);
            }
            if (el.pluralType === 'ordinal' &&
                blacklist.includes(Element.selectordinal)) {
                throw new BlacklistElement(Element.selectordinal);
            }
        }
        if (intl_messageformat_parser_1.isSelectElement(el) || intl_messageformat_parser_1.isPluralElement(el)) {
            const { options } = el;
            for (const selector of Object.keys(options)) {
                verifyAst(blacklist, options[selector].value);
            }
        }
    }
}
function checkNode(context, node, importedMacroVars) {
    const msgs = util_1.extractMessages(node, importedMacroVars);
    if (!msgs.length) {
        return;
    }
    const blacklist = context.options[0];
    if (!Array.isArray(blacklist) || !blacklist.length) {
        return;
    }
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst(context.options[0], intl_messageformat_parser_1.parse(defaultMessage));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e.message,
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow specific elements in ICU message format',
            category: 'Errors',
            recommended: false,
            url: 'https://github.com/formatjs/formatjs/tree/master/packages/eslint-plugin-formatjs#blacklist-elements',
        },
        fixable: 'code',
        schema: [
            {
                type: 'array',
                properties: {
                    items: {
                        type: 'string',
                        enum: Object.keys(Element),
                    },
                },
            },
        ],
    },
    create(context) {
        let importedMacroVars = [];
        return {
            ImportDeclaration: node => {
                const moduleName = node.source.value;
                if (moduleName === '@formatjs/macro' || moduleName === 'react-intl') {
                    importedMacroVars = context.getDeclaredVariables(node);
                }
            },
            JSXOpeningElement: (node) => checkNode(context, node, importedMacroVars),
            CallExpression: node => checkNode(context, node, importedMacroVars),
        };
    },
};
exports.default = rule;
