// eslint-plugin-formatjs/dist/rules/enforce-description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
function checkNode(context, node, importedMacroVars) {
    const msgs = util_1.extractMessages(node, importedMacroVars);
    for (const [{ message: { description }, descriptionNode, },] of msgs) {
        if (!description && !descriptionNode) {
            context.report({
                node,
                message: '`description` has to be specified in message descriptor',
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Enforce description in message descriptor',
            category: 'Errors',
            recommended: false,
            url: 'https://github.com/formatjs/formatjs/tree/master/packages/eslint-plugin-formatjs#enforce-description',
        },
        fixable: 'code',
    },
    create(context) {
        let importedMacroVars = [];
        return {
            ImportDeclaration: node => {
                const moduleName = node.source.value;
                if (moduleName === '@formatjs/macro' || moduleName === 'react-intl') {
                    importedMacroVars = context.getDeclaredVariables(node);
                }
            },
            JSXOpeningElement: (node) => checkNode(context, node, importedMacroVars),
            CallExpression: node => checkNode(context, node, importedMacroVars),
        };
    },
};
exports.default = rule;
