// eslint-plugin-formatjs/dist/rules/no-offset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
const intl_messageformat_parser_1 = require("intl-messageformat-parser");
class NoOffsetError extends Error {
    constructor() {
        super(...arguments);
        this.message = 'offset are not allowed in plural rules';
    }
}
function verifyAst(ast) {
    for (const el of ast) {
        if (intl_messageformat_parser_1.isPluralElement(el)) {
            if (el.offset) {
                throw new NoOffsetError();
            }
            const { options } = el;
            for (const selector of Object.keys(options)) {
                verifyAst(options[selector].value);
            }
        }
    }
}
function checkNode(context, node, importedMacroVars) {
    const msgs = util_1.extractMessages(node, importedMacroVars);
    for (const [{ message: { defaultMessage }, messageNode, },] of msgs) {
        if (!defaultMessage || !messageNode) {
            continue;
        }
        try {
            verifyAst(intl_messageformat_parser_1.parse(defaultMessage));
        }
        catch (e) {
            context.report({
                node: messageNode,
                message: e.message,
            });
        }
    }
}
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'Disallow offset in plural rules',
            category: 'Errors',
            recommended: false,
            url: 'https://github.com/formatjs/formatjs/tree/master/packages/eslint-plugin-formatjs#no-offset',
        },
        fixable: 'code',
    },
    create(context) {
        let importedMacroVars = [];
        return {
            ImportDeclaration: node => {
                const moduleName = node.source.value;
                if (moduleName === '@formatjs/macro' || moduleName === 'react-intl') {
                    importedMacroVars = context.getDeclaredVariables(node);
                }
            },
            JSXOpeningElement: (node) => checkNode(context, node, importedMacroVars),
            CallExpression: node => checkNode(context, node, importedMacroVars),
        };
    },
};
exports.default = rule;
