// @box/frontend/prettier/prettierrc.js
module.exports = {
    arrowParens: 'avoid',
    bracketSpacing: true,
    insertPragma: false,
    jsxBracketSameLine: false,
    jsxSingleQuote: false,
    printWidth: 120,
    quoteProps: 'as-needed',
    requirePragma: false,
    semi: true,
    singleQuote: true,
    tabWidth: 4,
    trailingComma: 'all',
    useTabs: false,
    overrides: [
        {
            files: ['*.ts', '*.tsx'],
            options: { parser: 'typescript' },
        },
        {
            files: '*.js',
            options: { parser: 'flow' },
        },
        {
            files: '*.md',
            options: {
                parser: 'markdown',
                tabWidth: 2,
                printWidth: 80,
            },
        },
        {
            files: '*.scss',
            options: {
                parser: 'scss',
                printWidth: 512,
            },
        },
        {
            files: '*.css',
            options: { parser: 'css' },
        },
        {
            files: '*.gql',
            options: { parser: 'graphql' },
        },
        {
            files: '*.html',
            options: { parser: 'html' },
        },
        {
            files: '*.json',
            options: { parser: 'json' },
        },
    ],
};
